build-tools/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java

```diff
@@ -153,17 +159,26 @@ private void createListClustersTask(Project project, NamedDomainObjectContainer<
 
     }
 
-    static class TestClustersHookPlugin implements Plugin<Project> {
-        @Override
+    static abstract class TestClustersHookPlugin implements Plugin<Project> {
+        @Inject
+        public abstract BuildEventsListenerRegistry getEventsListenerRegistry();
+
+        @Inject
+        public TestClustersHookPlugin() {}
+
         public void apply(Project project) {
             if (project != project.getRootProject()) {
                 throw new IllegalStateException(this.getClass().getName() + " can only be applied to the root project.");
             }
-
             Provider<TestClustersRegistry> registryProvider = GradleUtils.getBuildService(
                 project.getGradle().getSharedServices(),
                 REGISTRY_SERVICE_NAME
             );
+
+            Provider<TaskEventsService> testClusterTasksService = project.getGradle()
+                .getSharedServices()
```

Comment:
I'm wondering if we can ditch `getBuildService()` in favor of `@ServiceReference` in most places.

Reply:
@ServiceReference only works for tasks to detangle the registration of a build service in a plugin and its usage within tasks.